tests/test_single_topology.py

```diff
@@ -1691,42 +1614,85 @@ def _assert_u_and_grad_consistent(u_fwd, u_rev, x_fwd, x_rev, fused_map):
-        mols = read_sdf(path_to_ligand)
+    ff = Forcefield.load_default()
+    # map atoms in the combined mol_ab to the atoms in the combined mol_ba
+    fused_map = _get_fused_map(mol_a, mol_b, core)
+
+    st_fwd = SingleTopology(mol_a, mol_b, core, ff)
+    st_rev = SingleTopology(mol_b, mol_a, core[:, ::-1], ff)
+    conf_a = get_romol_conf(mol_a)
+    conf_b = get_romol_conf(mol_b)
+    test_conf = st_fwd.combine_confs(conf_a, conf_b, 0)
+
+    for lamb in np.linspace(0, 1, 12):
+        sys_fwd = st_fwd.setup_intermediate_state(lamb)
+        sys_rev = st_rev.setup_intermediate_state(1 - lamb)
+
+        _assert_u_and_grad_consistent(
+            sys_fwd.bond, sys_rev.bond, test_conf, fused_map, canon_fn=lambda idxs, _: tuple(canonicalize_bond(idxs))
+        )
 
-    for idx, mol_a in enumerate(mols):
-        for mol_b in mols[idx + 1 :]:
-            print("testing", mol_a.GetProp("_Name"), "->", mol_b.GetProp("_Name"))
-            core = atom_mapping.get_cores(mol_a, mol_b, **DEFAULT_ATOM_MAPPING_KWARGS)[0]
-            amm_fwd = AtomMapMixin(mol_a, mol_b, core)
-            amm_rev = AtomMapMixin(mol_b, mol_a, core[:, ::-1])
-            fused_map = np.concatenate(
-                [
-                    np.array([[x, y] for x, y in zip(amm_fwd.a_to_c, amm_rev.b_to_c)], dtype=np.int32).reshape(-1, 2),
-                    np.array(
-                        [
-                            [x, y]
-                            for x, y in zip(amm_fwd.b_to_c, amm_rev.a_to_c)
-                            if x not in core[:, 0] and y not in core[:, 1]
-                        ],
-                        dtype=np.int32,
-                    ).reshape(-1, 2),
-                ]
-            )
-            sys_fwd, conf_fwd = get_vacuum_system_and_conf(mol_a, mol_b, core, 0.0)
-            sys_rev, conf_rev = get_vacuum_system_and_conf(mol_b, mol_a, core[:, ::-1], 1.0)
-
-            _assert_u_and_grad_consistent(sys_fwd.bond, sys_rev.bond, conf_fwd, conf_rev, fused_map)
-            _assert_u_and_grad_consistent(sys_fwd.angle, sys_rev.angle, conf_fwd, conf_rev, fused_map)
-            _assert_u_and_grad_consistent(sys_fwd.nonbonded, sys_rev.nonbonded, conf_fwd, conf_rev, fused_map)
-            _assert_u_and_grad_consistent(sys_fwd.torsion, sys_rev.torsion, conf_fwd, conf_rev, fused_map)
-            _assert_u_and_grad_consistent(sys_fwd.chiral_atom, sys_rev.chiral_atom, conf_fwd, conf_rev, fused_map)
+        _assert_u_and_grad_consistent(
+            sys_fwd.angle, sys_rev.angle, test_conf, fused_map, canon_fn=lambda idxs, _: tuple(canonicalize_bond(idxs))
+        )
+
+        # for propers, we format the phase as a 5 decimal string to guard against loss of precision
+        _assert_u_and_grad_consistent(
+            sys_fwd.proper,
+            sys_rev.proper,
+            test_conf,
+            fused_map,
+            canon_fn=lambda idxs, params: tuple([*canonicalize_bond(idxs), f"{params[1]:.5f}", int(params[2])]),
```

Comment:
surprised to see this non-numeric , might use `np.round` or similar

Reply:
good call - I think `int()` is effectively a `floor()` - will fix.

Reply:
had to use int(round(params[2])) to go from int(round(jax.ArrayImpl) -> jax.ArrayImpl) -> int

Reply:
Ahh, sorry meant about the non-numeric type (suggesting `f"{params[1]:.5f}"` -> `np.round(params[1], 5)` or similar). This doesn't matter functionally (only used for equivalence-testing), converting this parameter to a non-numeric type is fine.